darkdetect/base.py

```diff
@@ -0,0 +1,89 @@
+from typing import Callable
+from enum import Enum, auto
+
+
+class ListenerState(Enum):
+    """
+    A listener state
+    """
+    Listening = auto()
+    Stopping = auto()
+    Dead = auto()
+
+
+class BaseListener:
+    """
+    An abstract listener class
+    Subclasses promise that it is safe to call stop() then wait()
+    from a different thread than listen() was called in; provides two
+    threads are not both racing to call these methods
+    """
+
+    def __init__(self, callback: Callable[[str], None]):
+        """
+        :param callback: The callback to use when the listener detects something
+        """
+        self._state: ListenerState = ListenerState.Dead
+        self.callback: Callable[[str], None] = callback
+
+    def listen(self):
+        """
+        Start the listener if it is not already running
+        """
+        if self._state == ListenerState.Listening:
+            raise RuntimeError("Do not run .listen() from multiple threads concurrently")
+        if self._state == ListenerState.Stopping:
+            raise RuntimeError("Call .wait() to wait for the previous listener to finish shutting down")
+        self._state = ListenerState.Listening
+        try:
+            self._listen()
+        except NotImplementedError:
+            self._state = ListenerState.Dead
+            raise
+        except Exception as e:
+            self.stop()  # Just in case
+            raise RuntimeError("Listen failed") from e
+
+    def stop(self):
+        """
+        Tells the listener to stop; may return before the listener has stopped
+        If the listener is not currently listening, this is a no-op
+        This function may be called if .listen() errors
+        """
+        if self._state == ListenerState.Listening:
+            self._stop()
+            self._state = ListenerState.Stopping
+
+    def wait(self):
```

Comment:
Something about the names `stop` and `wait` still confuses me a little. `wait` is particularly puzzling because it is not clear what we are waiting for, it might be something related to the listening itself. 

Perhaps we need to change these to a different set of names, such as `kill` and `stop`? Otherwise, I would be fine with more explicit names, e.g. `kill_immediately` and `stop_and_wait_for_termination`.

Reply:
Hmm, I was going based off of subprocess names somewhat but preferred `stop` to `kill` since I wasn't sure how the Windows `stop` function would work. I'm inclined to say docstrings and documentation here is sufficient, since I'm not sure what good names would be for these.

These are intended to be use in different threads than the listen function for obvious reasons. I avoided `join` because I didn't want to confuse this with actually managing threads. I'm open to suggestions, but the most accurate names I could think of might be `request_stop_no_block` and `request_and_wait_for_stop`; but these seem clunky. Ultimately this is why I went with `stop` and `wait`, but I'm not set on those names.

Reply:
IMHO, `wait` is not a valid name for this functionality. As I said, I am fine with a more descriptive name, otherwise I would use `stop` for this method and rename the former `stop` to any of the alternative names such as kill, halt, terminate, interrupt, …

Reply:
Hmm, how about: `stop -> initiate_shutdown` and `wait -> initiate_and_wait_for_shutdown`

We could also unify the functions into: `def stop(blocking: bool, timeout: Optional[int] = None)` too; where `stop -> stop(False)` and `wait -> stop(True)`; I would be against a default argument in this case, i.e. require the user to specify the boolean value.

---

#### My Preference
Actually, we could simply unify this as `def stop(timeout: Optional[int])`, where: `timeout = 0` has no timeout, and `None` means infinite. Then instead of raising an exception on timeout, we return `True` if the listener has finished stopping and `False` if there was a timeout. No default argument means the user is forced to be conscious of this. I think I like this idea honestly. Do you find any issues with this @albertosottile ?

---

My main hangup with `wait -> stop` (and `stop -> initiate_stop`) is the Windows listener really. It'd be ok if `stop()` took a few seconds, but the Windows listener might never actually stop, so I think it'd be easy to accidentally assume stop (`wait`) returns in quickly when phrases just as `stop`. Obviously, this could be mitigated by making the Windows stop function actually stop. A trivial way to do this is to shove it in another process like the others, but there is likely a better way, like what you mentioned above. If the Windows stopper / wait function was relatively quick (seconds rather than undefined length), I would be ok with: `stop -> initiate_stop` and `wait -> stop`. Or something like that.

Reply:
> Actually, we could simply unify this as def stop(timeout: Optional[int]), where: timeout = 0 has no timeout, and None means infinite. Then instead of raising an exception on timeout, we return True if the listener has finished stopping and False if there was a timeout. No default argument means the user is forced to be conscious of this. I think I like this idea honestly. This would also simplify the API and make things cleaner I think. Do you find any issues with this @albertosottile ?

I like this, but I need some clarification on the effects of the argument:
 
- `timeout = None`: no timeout, we wait indefinitely until the listener stops. What will happen on Windows?
- `timeout = 3` or any other number: we wait until the listener stops for 3 seconds, after the 3 seconds we return as described above depending on the final listener status
- `timeout = 0`: no wait, return happens immediately (and I guess it is always `False`?)

Do I understand this correctly?

Re the Windows listener: I plan to look at it and `RegCloseKey` after this is merged, likely after the holidays.

Reply:
> * `timeout = None`: no timeout, we wait indefinitely until the listener stops. What will happen on Windows?
Yes.
On Windows (until you look at `RegCloseKey`), we would:
1. Initiate the stop: This means that `self.callback` would no longer be called because of the `if` gate within the `_listen` function
2. Hang until the theme is updated again
3. The `_listen` would call the callback, which would notice the if gate, no-op, then both would return
4. `stop(None)` would exit successfully

> * `timeout = 3` or any other number: we wait until the listener stops for 3 seconds, after the 3 seconds we return as described above depending on the final listener status

Correct

> * `timeout = 0`: no wait, return happens immediately (and I guess it is always `False`?)

Yes. It technically doesn't _have_ to return False if you can immediately remove the listener and assert all callbacks have completed running, but for the three present listeners, it would.

> Re the Windows listener: I plan to look at it and `RegCloseKey` after this is merged, likely after the holidays.

👍

Reply:
I agree with everything except

> Hang until the theme is updated again

I do not think we can allow this, as this is basically a limitation of our implementation. I would rather state in the README that, for the moment, `stop(timeout=None)` is not supported on Windows (and raise `ValueError` with an explanation, if called). I will then investigate how to fix this in a subsequent PR.

Reply:
Hmm, I suppose a motivated user could still emulate `None` with a simple loop so this isn't a loss in functionality, only an API choice. Sounds good.